src/main/scala/io/iohk/ethereum/network/rlpx/MessageCodec.scala

```diff
@@ -33,7 +37,16 @@ class MessageCodec(frameCodec: FrameCodec, messageDecoder: MessageDecoder, proto
     }
   }
 
-  // TODO: ETCM-402 - messageDecoder should use negotiated protocol version
+  private def setRemoteProtocolVersionBasedOnHelloMessage(m: Message): Unit = {
+    if (remotePeerProtocolVersion.isEmpty) {
+      m match {
+        case hello: Hello =>
+          remotePeerProtocolVersion = protocolNegotiator.negotiate(hello.capabilities)
```

Comment:
wdyt of `protocolNegotiator` per connections, and keeping there state about p2pVersion and protocol version in some thread safe manner ?  ( just do not like that now we call `protocolNegotiator.negotiate(hello.capabilities)` in two places with same argument, and if some in the future will look only on `EtcHelloExchangeState` that this negotiation also happens on layer bellow for decoding purposes)

Reply:
it something very similar to besu implementation - IMHO it's a good idea, but not for this PR, but for new task :) I will create one

Reply:
https://jira.iohk.io/browse/ETCM-474